packages/core/shared.js

```diff
@@ -957,4 +960,178 @@ function checkTmOptionalPointer(td){
+                Status: 'not-impl',
+                Comment: 'td does not link to tm or links to more than one tm'
+            },
+            {
+                ID: ASSERTION_OPTIONAL,
+                Status: 'not-impl',
+                Comment: 'td does not link to tm or links to more than one tm'
+            }
+        ];
+    }
+    typeLink = typeLink[0];
+
+    const fileFetcher = async (url) => {
+        try {
+            return {
+                'success': true,
+                'data': JSON.parse(fs.readFileSync(url.split('file://')[1]))
+            };
+        } catch {
+            return {
+                'success': false,
+                'error': 'make sure you are not using file:// links inside non-browser environment'
+            };
+        }
+    }
+
+    const httpFetcher = async (url) => {
+        try {
+            return {
+                'success': true,
+                'data': await (await fetch(url)).json()
+            };
+        } catch {
+            return {
+                'success': false,
+                'error': 'make sure related tm is valid JSON and is not under CORS'
+            };
+        }
+    }
+
+    // TODO: Add support for other network protocols, e.g., MQTT
+    const fetcher = (typeLink.href.startsWith('file://')) ? fileFetcher : httpFetcher;
+    const result = await fetcher(typeLink.href);
+
+    if (!result.success) {
+        return [
+            {
+                ID: ASSERTION_REQUIRED,
+                Status: 'warning',
+                Comment: result.error
+            },
+            {
+                ID: ASSERTION_OPTIONAL,
+                Status: 'warning',
+                Comment: result.error
+            }
+        ];
+    }
+
+    const tm = result.data;
+    const tmAffordances = {};
+    for (const affordanceType of ['properties', 'actions', 'events']) {
+        tmAffordances[affordanceType] = {
+            all: Object.keys(tm[affordanceType] || {}),
+            optional: (tm['tm:optional'] || []).map(e => {
+                const x = e.split('/');
+                if (x[1] === affordanceType) return x[2];
+                return null;
+            }).filter(e => e),
+        };
+    }
+
+    // Check if arr2 is subset of arr1,
+    // i.e., all elements of arr2 are contained in arr1
+    const isSubset = (arr1, arr2) => arr2.every(e => arr1.includes(e));
+
+    // Check if arr1 and arr2 have any intersection
+    const haveIntersection = (arr1, arr2) => arr1.some(e => arr2.includes(e));
+
+    const results = [];
+    const requiredAssertion = false;
+    const optionalAssertion = false;
+    for (const affordanceType of ['properties', 'actions', 'events']) {
+        // Combine all and optional into one => required
+        const required = tmAffordances[affordanceType].all.filter(
+            e => !tmAffordances[affordanceType].optional.includes(e));
+
+        if (!isSubset(Object.keys(td[affordanceType]), required) && !requiredAssertion) {
```

Comment:
Here I get:

       

>  if (!isSubset(Object.keys(td[affordanceType]), required) && !requiredAssertion) {
>                              ^
> 
> TypeError: Cannot convert undefined or null to object
>     at Function.keys (<anonymous>)
>     at Object.checkLinkedAffordances (/home/eko/git/thingweb-playground/packages/core/shared.js:1087:30)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async /home/eko/git/thingweb-playground/packages/core/index.js:178:59
> 
> Node.js v19.2.0

The object keys actual get logged correctly when I put add console.log statements so I am not sure where the error comes from